laborator/lab-02/README.md

```diff
@@ -0,0 +1,554 @@
+
+Scăderea a doi pointeri este posibilă doar dacă ambii au același tip. Rezultatul scăderii este obținut prin calcularea diferenței adreselor de memorie către care pointează. Spre exemplu, calcularea lungimii unui șir de caractere:
+
+```c
+char *s = "Learn IOCLA, you must!";
+char *p = s;
+for (; *p; ++p); /* Se iterează caracter cu caracter, până la '\0' */
+ 
+printf("%ld", p - s); /* Se va afișa 22. */
+```
+#### **Interpretarea datelor din memorie**
+
+```c
+int n = 0xCAFEBABE;
+unsigned char first_byte = *((unsigned char*) &n); /* Se extrage primul byte al lui n */
+unsigned char second_byte = *((unsigned char*) &n + 1); /* Se extrage al doilea byte al lui n */
+printf("0x%x, 0x%x\n", first_byte, second_byte); /* Se va afișa 0xBE, 0xBA */
+```
+
+> **NOTE:** În cazul pointerilor castați, operațiile aritmetice se realizează pe tipul la care aceștia au fost castați.
+
+> **WARNING:** Nu faceți confuzie intre **\*p++** și **(\*p)++**. În primul caz se incrementează adresa la care pointează p, iar în cel de al doilea, valoarea de la acea adresă.
+> Aritmetica pe pointeri de tip void nu este posibilă din lipsa unui tip de date concret la care pointează.
+
+### **Pointeri la tablouri**
+
+Între pointeri și tablouri există o legatură foarte stransă. În C numele unui tablou este un *pointer constant* (adresa sa este alocată de către compilator și nu mai poate fi modificată în timpul execuției) la primul element din tablou: **v = &v[0]**. De exemplu:
+
+```c
+int v[10], *p;
+p = v;
+++p; /* Corect */
+++v; /* EROARE */
+```
+Vectorii sunt stocați într-o zonă continuă de memorie, astfel că aritmetica pe pointeri funcționează la fel și in cazul lor, obținând urmatoarele echivalențe:
+
+```c
+v[0] <==> *v
+v[1] <==> *(v + 1)
+v[n] <==> *(v + n)
+&v[0] <==> v
+&v[1] <==> v + 1
+&v[n] <==> v + n
+```
+
+De asemenea, un vector conţine şi informaţii legate de lungimea vectorului şi dimensiunea totală ocupată în memorie, astfel că operatorul **sizeof(v)** va returna spațiul ocupat în memorie (numărul de octeți), iar **sizeof(v) / sizeof(\*v)** va returna numărul de elemente ale lui **v**.
+
+Folosind pointeri putem să alocăm dinamic memorie. În acest sens, alocarea dinamică a unui tablou bidimensional (o matrice) se poate realiza astfel:
+
+Metoda tradițională, în care alocăm un array de pointeri la pointeri:
+
+```c
+int **array1 = malloc(nrows * sizeof(*array1));
+for (i = 0; i < nrows; ++i)
+    array1[i] = malloc(ncolumns * sizeof(**array1));
+```
+
+Dacă dorim sa păstrăm array-ul într-o zonă continuă de memorie:
+
+```c
+int **array2 = malloc(nrows * sizeof(*array2));
+array2[0] = malloc(nrows * ncolumns * sizeof(**array2));
+for (i = 1; i < nrows; ++i)
+    array2[i] = array2[0] + i * ncolumns;
+```
+
+Mai jos este prezentată diferența dintre cele doua abordări:
+
+![](https://ocw.cs.pub.ro/courses/_media/iocla/laboratoare/array1.gif)
+![](https://ocw.cs.pub.ro/courses/_media/iocla/laboratoare/array2.gif)
```

Comment:
Eu folosesc dark theme în github și imaginile se văd puțin ciudat pentru că nu au absolut deloc margini. Având în vedere că sunt două consecutive, poate ar fi bine să fie o poză cum ambele, care are și niște margini. Alternativ, ar fi bine să fie scos fundaul (deși nu o să se mai vadă aproape deloc cu dark theme...).

Reply:
Dacă modific imaginile unde ar trebui să le încarc?

Reply:
Asta e una din problemele pe care trebuie să le adresăm cumva pentru că dacă se schimbă ordinea laboratoarelor, o să devină invalide link-urile. Daniel le-a încărcat pe GitHub.

Reply:
Ar trebui sa avem un director images/ pentru fiecare laborator si sa punem imaginile acolo. Nu e in filozofia git sa incarcam binare insa nu prea avem alta optiune.

Reply:
Asta ar fi soluția și din punctul meu de vedere. @dbaluta  după ce facem și PR-ul acesta o să schimb structura directoarelor și o să facem symlink-uri către un alt director. Poate ar fi o soluție să facem asta și pe ocw, dar încă nu am luat decizia, deci pentru PR soluția este un director.

Reply:
@PinkiePie1189 vezi #30.